src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_internal.clj

```diff
@@ -86,6 +86,29 @@
    pool :- jruby-schemas/pool-queue-type]
   (.borrowItemWithTimeout pool timeout TimeUnit/MILLISECONDS))
 
+(schema/defn insert-shutdown-poison-pill
+  [pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (ShutdownPoisonPill. pool)))
+
+(schema/defn insert-poison-pill
+  [error :- Throwable
+   pool :- jruby-schemas/pool-queue-type]
+  (.insertPill pool (PoisonPill. error)))
+
+(schema/defn fill-pool-with-poison-pills!
+  "Fill an empty pool with poison pills. Should always be run on the
+  modify-instance-agent. insert-fn should be a fn that inserts a poison
+  pill of some kind into the pool"
+  [pool-state :- jruby-schemas/PoolState
+   insert-fn :- IFn]
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)]
+    (when (not (= pool-size (.remainingCapacity pool)))
```

Comment:
> Oh, in that case, I think it's 💯 to throw an Exception since the lock will never be available anyway.

Yeah, exactly.  I think this is ultimately going to be an improvement.  I think today that if we have a situation where a pill is inserted into the pool before it has been initially filled up or around shutdown time and a file sync lock is attempted, the `lock()` call will probably be stuck indefinitely - presumably until the process is SIGKILLed by the OS service manager.  I was mostly just trying to gauge whether going with an exception vs. just granting the lock was preferable.  I think an exception is better, though.  If we know we want to shut down as soon as possible after this happens, I wouldn't think we'd want to allow a potentially long-running background task like a file sync code update to be kicked off.

In terms of what kind of exception we'd throw here, I think it would be good to have an identifiable one that receivers could implement custom handling for.  For example, in the case of the exception being thrown because Trapperkeeper has called stop twice on the JRuby service, it might be good to just eat a 'poison pill' exception rather than just put extra noise into the logs - since, per the point @jpinsonault made elsewhere - we should log statements around where we do `insertPill` calls anyway about why the pill is being inserted.  It might make sense to codify a new `PillException` class and have the actual pill be accessible through a getter method on the exception in the event that the caller would like to / need to inspect it for some reason.

For convenience, it might make sense to just have a `PillException` be a class extended from `InterruptedException`.  That way, callers that don't care about the difference between the `InterruptedException` - which many of the methods in the `JRubyPool` class already throw - and a `PillException` could handle both types of exception as the base type.  Semantically, I think one could argue that `PillException` is a specialized form of interruption which can occur if a pill is inserted into the pool while the caller is waiting for a borrow, lock, or whatever to complete.

Definitely open to other names / suggestions around the naming, though.